app/services/update_location.rb

```diff
@@ -6,30 +6,45 @@ def initialize(location:, user:)
     @user = user
   end
 
-  def update!(params)
-    params = params.with_indifferent_access
-    if params['address']
-      params['address_id'] = Address.find_or_create_from_params(params.delete('address')).id
+  def update(params)
+    address = params.delete('address')
+    if address
+      address_params = build_address(address)
+      params.merge!(address_params)
     end
 
     return location if location.matches_params?(params)
 
     Location.transaction do
-      location.update_attributes!(state: 'old')
-      create_new_version!(params)
+      begin
+        location.update_attributes!(state: 'old')
+        create_new_version(params)
+      rescue ActiveRecord::RecordInvalid
```

Comment:
The process creates a new location rather than modifying an existing one, as such I need the update call to return the location so it can be used to show the errors. 

I tried to get away from raising the error in the first place but couldn't find a nice way to do it and still have the process be transactional.